src/threading.c

```diff
@@ -334,6 +334,37 @@ static jl_value_t *ti_run_fun(jl_callptr_t fptr, jl_method_instance_t *mfunc,
 jl_mutex_t codegen_lock;
 jl_mutex_t typecache_lock;
 
+// Init all RCU counters to zero
+int rcu_gp_ctr = 0;
+int* rcu_reader_gp;
+
+JL_DLLEXPORT void rcu_read_lock(void)
+{
+    rcu_reader_gp[jl_threadid()] = rcu_gp_ctr + 1;
```

Comment:
Also I'm pretty sure these needs to have the right ordering too. The unlock should have a release ordering. The lock must not be reordered with a read later so I feel like you actually need seq_cst here.